### PR TITLE
URL match improvements

### PR DESCRIPTION
**Pull request by SlavaSkvortsov**
Refactored URL matching, moved URL-related logic to a separate class. 
The main goal was to do a params-independent URL comparing. It was hard to use the lib for GET requests (had to use regex matching)

### CODE REVIEW

**Comment by lundberg** [comment]
Thanks for contributing to RESPX @SlavaSkvortsov !

I like the idea of breaking up the code for readability and easier (re)use.

Though, I think we need to take the discussion in #81, by @jocke-l, into account and start using `httpx.URL`.

Two reasons, less "duplicated" code from HTTPX to maintain, and to align with the HTTPX api, as discussed in #79, i.e. support `params` kwarg.

I'd like to see an api where one could do...

```py
@respx.mock(base_url="https://example.org")
async def test_something(respx_mock):
    respx_mock.post("/foobar/", params={"foo": "bar"})
```

...but also, *if possible*, even...

```py
@respx.mock(base_url="https://example.org")
async def test_something(respx_mock):
    respx_mock.post(r"/foobar/(?P<slug>\w+)/", params={"foo": "bar"})
```

Meaning, it would be nice to support both `base_url`, regular string `url`,  regex `url` and optional `params` for merge.

**Comment by SlavaSkvortsov** [comment]
Thank you for maintaining this lib, you are doing a great job! 

I'll add `params` support and use `httpx.URL`, but I'm strongly against of idea comparing params as strings (or, even more, as a part of [regex pattern](https://github.com/lundberg/respx/pull/81/files#diff-33abcd067384ca826af1fb9f361b6182R282)). Do you see any profit in string-based comparing or it will be fine to compare them as dicts?

**Comment by lundberg** [comment]
Thanks for the update, but I think you mis-interpreted by last comment ;-)

What I meant was that it would be best if this PR extends/depends on #81.

It's better to keep each PR small and specific to make reviews easier and to minimize introduction of bugs. Also the commits and changelog gets clearer this way.

Let's merge #81 first, and then come back to this PR, sounds good @SlavaSkvortsov ?

**Comment by SlavaSkvortsov** [comment]
Yeap, sounds great :)

**Comment by SlavaSkvortsov** [comment]
This one looks ready as well. 
The main idea here - you should never compare URL params as strings (or add them as a part of regex). Also, if your pattern is `http://foo.bar` the URL `http://foo.bar?qwe=qwe` should match.

File: respx/models.py [hunk 1]
```diff
@@ -379,21 +354,92 @@ def match(self, request: Request) -> Optional[Union[Request, ResponseTemplate]]:
...
+            base_url = httpx.URL(base if base.endswith("/") else base + "/")
+
+            if not url:
+                url = base_url
+
+        if params is not None:
+            self._params = httpx.QueryParams(params)
+
+        if isinstance(url, (str, tuple, httpx.URL)):
+            self._url = base_url.join(httpx.URL(url))
```
Comment (lundberg) [inline respx/models.py:394]: Here will `base_url` be joined with it self, when `url` arg is `None`, due to line `388`.

Maybe not a problem?

**Review by lundberg — COMMENTED** [review]
Nice rebase and refactoring!

File: respx/models.py [hunk 2]
```diff
@@ -379,21 +354,92 @@ def match(self, request: Request) -> Optional[Union[Request, ResponseTemplate]]:
...
+                "Request url pattern must be str or compiled regex, got {}.".format(
+                    type(url).__name__
+                )
+            )
+
+    def matches(self, url_parts: URL) -> Tuple[bool, Dict[str, Any]]:
+        url = httpx.URL(url_parts)
+        url, params = self._extract_params(url)
+
+        if self._params is not None and self._params != params:
```
[the remark this answers is not in the document]
Comment (SlavaSkvortsov) [reply respx/models.py:427]: > Also, should we require full match on all params, or treat them as pattern matching and accept kwargs params only to be included in possibly "richer" request url params?

It's a nice question. I thought about it today. I think the second option is better

File: respx/models.py (hunk 1 above)
Comment (SlavaSkvortsov) [reply respx/models.py:394]: Oh, I'll take a look

File: respx/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lundberg) [reply respx/models.py]: I'm also leaning to second option, i.e. accept pattern params to be subset of request params.

Major reason is that it aligns with your proposal of empty/missing pattern params matching requests with params.

**Comment by lundberg** [comment]
Sorry, but this one is too outdated now as of #96. 

Please check out the new patterns to see if any of your logic is missing.

**Pull request closed without merging**